File: algorithms/trees/findMinForBinarySearchTree.js
class Node {
  constructor(x) {
    this.data = x;
    this.left = null;
    this.right = null;
  }
}

class BinarySearchTree {
  constructor() {
    this.root = null;

    this.insert = this.insert.bind(this);
    this.search = this.search.bind(this);
    this.findMin = this.findMin.bind(this);
  }

  insert(root, x) {
    if (root == null) {
      root = new Node(x);
    } else if (x <= root.data) {
      root.left = this.insert(root.left, x);
    } else {
      root.right = this.insert(root.right, x);
    }
    return root;
  }

  search(root, x) {
    if (root == null) {
      return false;
    } else if (x == root.data) {
      return true;
    } else if (x <= root.data) {
      return this.search(root.left, x);
    } else {
      return this.search(root.right, x);
    }
  }

  findMin(root) {
    if (root == null) {
      return console.log('Tree is empty.');
    } else if (root.left == null) {
      return root.data;
    } else {
      return this.findMin(root.left);
    }
  }
}

console.log('Starting job...');
var myTree = new BinarySearchTree();
console.log('The min is: ' + myTree.findMin(myTree.root));
var root = myTree.root;
root = myTree.insert(root, 2);
myTree.root = root;
root = myTree.insert(root, 4);
root = myTree.insert(root, 10);
root = myTree.insert(root, 6);
root = myTree.insert(root, 8);
console.log('Search for 8: ' + myTree.search(myTree.root, 8));
console.log('Search for 11: ' + myTree.search(myTree.root, 11));
console.log('Search for 9: ' + myTree.search(myTree.root, 9));
console.log('Search for 10: ' + myTree.search(myTree.root, 10));
console.log('The min is: ' + myTree.findMin(myTree.root));

File: algorithms/trees/findMinForBinarySearchTree.java
class MinBinarySearchTree {
  class Node {
    int data;
    Node left;
    Node right;

    public Node(int x) {
      this.data = x;
      this.left = null;
      this.right = null;
    }
  }

  Node root;

  public MinBinarySearchTree() {
    this.root = null;
  }

  public Node insert(Node root, int data) {
    if (root == null) {
      root = new Node(data);
    } else if (data <= root.data) {
      root.left = insert(root.left, data);
    } else {
      root.right = insert(root.right, data);
    }
    return root;
  }

  public boolean search(Node root, int data) {
    if (root == null) {
      return false;
    } else if (data == root.data) {
      return true;
    } else if (data <= root.data) {
      return search(root.left, data);
    } else {
      return search(root.right, data);
    }
  }

  public int findMin(Node root) {
    if (root == null) {
      System.out.println("Tree is empty.");
      return -1;
    } else if (root.left == null) {
      return root.data;
    } else {
      return findMin(root.left);
    }
  }

  public static void main(String[] args) {
    MinBinarySearchTree myTree = new MinBinarySearchTree();
    Node root = myTree.root;
    System.out.println("Search for 8: " + myTree.search(null, 8));
    System.out.println("The min is: " + myTree.findMin(myTree.root));
    root = myTree.insert(root, 2);
    myTree.root = root;
    root = myTree.insert(root, 10);
    root = myTree.insert(root, 4);
    root = myTree.insert(root, 8);
    root = myTree.insert(root, 6);
    root = myTree.insert(root, 12);
    System.out.println("Search for 8: " + myTree.search(myTree.root, 8));
    System.out.println("Search for 12: " + myTree.search(myTree.root, 12));
    System.out.println("Search for 7: " + myTree.search(myTree.root, 7));
    System.out.println("The min is: " + myTree.findMin(myTree.root));
  }
}